vscode/quint-vscode/server/src/complete.ts

```diff
@@ -0,0 +1,250 @@
+ * Find a declaration of `name` by name inside `decl`.
+ */
+function findDeclByNameInDecl(name: string, decl: QuintDeclaration): bigint[] {
+  const visitor = new DeclOfNameFinder(name)
+  walkDeclaration(visitor, decl)
+  return visitor.namesFound
+}
+/**
+ * Find a declaration of `name` by name inside `expr`.
+ */
+function findDeclByNameInExpr(name: string, expr: QuintEx): bigint[] {
+  const visitor = new DeclOfNameFinder(name)
+  walkExpression(visitor, expr)
+  return visitor.namesFound
+}
+class DeclOfNameFinder implements IRVisitor {
+  nameToFind: string
+  namesFound: bigint[]
+
+  constructor(nameToFind: string) {
+    this.nameToFind = nameToFind
+    this.namesFound = []
+  }
+
+  /** Definitions **/
+  exitDef(def: QuintDef) {
+    if (def.name == this.nameToFind) {
+      this.namesFound.push(def.id)
+    }
+  }
+
+  /** Expressions **/
+  exitLambda(lambda: QuintLambda) {
+    lambda.params.forEach(param => {
+      if (param.name == this.nameToFind) {
+        this.namesFound.push(param.id)
+      }
+    })
+  }
+}
+
+/**
+ * Search `modules` for a declaration of `name` inside the module, definition, or expression identified by `scope`.
+ *
+ * If `scopeId` identifies a definition or expression, search it recursively.
+ * If `scopeId` identifies a module, search the module's top-level variables, but
+ * don't descend into any of its declarations.
+ *
+ * @param name Find a declaration of this name.
+ * @param scopeId ID of the module, definition, or expression to search.
+ * @param modules Array of modules that contain `scope`.
+ *
+ * @returns Array of IDs that declare `name`.
+ */
+function findDeclByNameInsideScope(name: string, scopeId: bigint, modules: QuintModule[]): bigint[] {
+  const module = modules.find(module => module.id == scopeId)
+  if (module) {
+    // scope is a module; return all variables with name `name`
+    return module.declarations.filter(decl => decl.kind == 'var' && decl.name == name).map(decl => decl.id)
+  }
+  const expr = findExpressionWithId(modules, scopeId)
+  if (expr) {
+    // scope is an expression; recursively search it for declarations of the name `name`
+    return findDeclByNameInExpr(name, expr)
+  }
+  const def = findDefinitionWithId(modules, scopeId)
+  if (def) {
+    // scope is a declaration; recursively search it for declarations of the name `name`
+    return findDeclByNameInDecl(name, def)
+  }
+  return []
+}
+
+/**
+ * Search for a declaration of `name`, in increasingly wide effects around `pos`.
+ *
+ * @returns Array of IDs that declare `name`.
+ */
+function findDeclByNameAtPos(name: string, pos: Position, sourceFile: string, parsedData: ParserPhase3): bigint[] {
+  // Compute effects that contain the triggering position, from the smallest to the largest.
+  const { modules, sourceMap } = parsedData
+  const results: [bigint, null][] = [...sourceMap.keys()].map(id => [id, null])
+  const effectsContainingPosition = findMatchingResults(sourceMap, results, pos, sourceFile)
```

Comment:
Ah, I think I was confused early on, because the comments in [`resultsOnPosition()`](https://github.com/informalsystems/quint/blob/5a36609ceb45178f6dd7069acdf45f9b1337c2c4/vscode/quint-vscode/server/src/reporting.ts#L145) refer to effects.

Maybe we should clean it up there as well?